### PR TITLE
[main] Add mina as a git submodule for o1js

### PR DESCRIPTION
**Pull request by MartinMinkov**
# Description 
Addresses https://github.com/o1-labs/o1js/issues/1096

:link: Mina: https://github.com/MinaProtocol/mina/pull/14461
:link: o1js-bindings: https://github.com/o1-labs/o1js-bindings/pull/159

This PR aims to add mina as a submodule of o1js and add support to build everything in the top level directory. To do so, the following changes were made:
1. Add mina as a submodule, located under `src/mina`, and target the PR above
2. Update the bindings submodule to the PR above
3. Add details to `README-dev.md` to elaborate on how to build the bindings and give contextual information
4. Some auxiliary updates, like updating `package-lock.json`, modifying some config files, etc.
5. Add a top-level `dune-project` file to let dune build all OCaml dependencies in the current root context.

### CODE REVIEW

File: README-dev.md [hunk 1]
```diff
@@ -1,80 +1,134 @@
...
+
+## Prerequisites
+
+Before starting, ensure you have the following tools installed:
 
-./run src/examples/api_exploration.ts
+- [Git](https://git-scm.com/)
+- [Node.js and npm](https://nodejs.org/)
+- [Dune](https://github.com/ocaml/dune)
+- [Cargo](https://www.rust-lang.org/learn/get-started)
```
Comment (mitschabaude) [inline README-dev.md:17]: I don't want to scare JS devs away from contributing -- maybe you could add a qualification here, similar to this:
```suggestion
- [Dune](https://github.com/ocaml/dune) (only needed when compiling o1js from source)
- [Cargo](https://www.rust-lang.org/learn/get-started) (only needed when compiling o1js from source)
```

File: README-dev.md [hunk 2]
```diff
@@ -1,80 +1,134 @@
...
+
+## Building Bindings
+
+If you need to regenerate the OCaml and WebAssembly artifacts, you can do so within the o1js repo. The [bindings](https://github.com/o1-labs/o1js-bindings) and [Mina](https://github.com/MinaProtocol/mina) repos are both submodules of o1js, so you can build them from within the o1js repo.
 
-./run-in-browser.js src/examples/api_exploration.ts
+o1js depends on OCaml code that is transplied to JavaScript using [Js_of_ocaml](https://github.com/ocsigen/js_of_ocaml), and Rust code that is transpiled to WebAssembly using [wasm-pack](https://github.com/rustwasm/wasm-pack). These artifacts allow o1js to call into [Pickles](https://github.com/o1-labs/snarkyhttps://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/README.md), [snarky](https://github.com/o1-labs/snarky), and [Kimchi](https://github.com/o1-labs/proof-systems) to write zk-SNARKs and zkApps.
```
Comment (mitschabaude) [inline README-dev.md:40]: typo
```suggestion
o1js depends on OCaml code that is transpiled to JavaScript using [Js_of_ocaml](https://github.com/ocsigen/js_of_ocaml), and Rust code that is transpiled to WebAssembly using [wasm-pack](https://github.com/rustwasm/wasm-pack). These artifacts allow o1js to call into [Pickles](https://github.com/o1-labs/snarkyhttps://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/README.md), [snarky](https://github.com/o1-labs/snarky), and [Kimchi](https://github.com/o1-labs/proof-systems) to write zk-SNARKs and zkApps.
```

File: README-dev.md [hunk 3]
```diff
@@ -1,80 +1,134 @@
...
-  npm run test:e2e
-  npm run e2e:show-report
-  ```
+For the wasm build, the output files are:
 
-## Branch Compatibility
+- `plonk_wasm_bg.wasm.d.ts`: TypeScript definition files describing the types of .wasm or .js files.
+- `plonk_wasm.d.ts`: TypeScript definition file for plonk_wasm.js.
+- `plonk_wasm_bg.wasm`: The compiled WebAssembly binary.
+- `plonk_wasm.js`: JavaScript file that wraps the WASM code for use in Node.js.
```
Comment (mitschabaude) [inline README-dev.md:73]: nit: the order of these is a bit weird
```suggestion
- `plonk_wasm_bg.wasm`: The compiled WebAssembly binary.
- `plonk_wasm_bg.wasm.d.ts`: TypeScript definition files describing the types of .wasm or .js files.
- `plonk_wasm.js`: JavaScript file that wraps the WASM code for use in Node.js.
- `plonk_wasm.d.ts`: TypeScript definition file for plonk_wasm.js.
```

Comment (mitschabaude) [reply README-dev.md:73]: also, isn't it simply:
```suggestion
- `plonk_wasm_bg.wasm.d.ts`: TypeScript definition files describing the types of `plonk_wasm_bg.wasm`.
- `plonk_wasm.d.ts`: TypeScript definition file for plonk_wasm.js.
- `plonk_wasm_bg.wasm`: The compiled WebAssembly binary.
- `plonk_wasm.js`: JavaScript file that wraps the WASM code for use in Node.js.
```

File: package.json [hunk 4]
```diff
@@ -43,11 +43,11 @@
   },
   "scripts": {
     "dev": "npx tsc -p tsconfig.node.json && node src/build/copy-to-dist.js",
-    "make": "make -C ../../.. snarkyjs",
     "make:no-types": "npm run clean && make -C ../../.. snarkyjs_no_types",
     "wasm": "./src/bindings/scripts/update-wasm-and-types.sh",
     "bindings": "cd ../../.. && ./scripts/update-snarkyjs-bindings.sh && cd src/lib/snarkyjs",
     "build": "node src/build/copy-artifacts.js && rimraf ./dist/node && npm run dev && node src/build/buildNode.js",
+    "build:bindings": "./src/bindings/scripts/update-snarkyjs-bindings.sh",
```
Comment (mitschabaude) [inline package.json:50]: I like `build:bindings`! but the script it replaces is `bindings`, not `make`. I still would like to have (a version of) `make` to build from source without changing files in the git source tree

**Review by mitschabaude — COMMENTED** [review]
Very nice how you updated the dev README!

File: package.json (hunk 4 above)
Comment (MartinMinkov) [reply package.json]: Fair enough! I added different script for this purpose, lemme know if it works for this case!

**Review by mitschabaude — APPROVED** [review]
Pre-approving with a few more comments.

Tried it out locally, it works for me. Let's go 🚀

**Pull request merged**